homeassistant/components/thethingsnetwork/coordinator.py

```diff
@@ -0,0 +1,149 @@
+
+            # Register newly found entities - nop if no new entities
+            self.__async_add_entities(measurements)
+
+            # Return measurements
+            return measurements
+        except TTNAuthError as err:
+            # Raising ConfigEntryAuthFailed will cancel future updates
+            # and start a config flow with SOURCE_REAUTH (async_step_reauth)
+            _LOGGER.error("TTNAuthError")
+            raise ConfigEntryAuthFailed from err
+        except Exception as err:
+            _LOGGER.error(err)
+            _LOGGER.error(traceback.format_exc())
+            raise
+
+    async def _push_callback(self, data: TTNClient.DATA_TYPE) -> None:
+        _LOGGER.debug("pushed data: %s", data)
+
+        # Register newly found entities - nop if no new entities
+        self.__async_add_entities(data)
+
+        # Push data to entities
+        self.async_set_updated_data(data)
+
+    def register_platform_entity_class(
+        self,
+        entity_class: type["TTN_Entity"],
+        async_add_entities: Callable[[TTNClient.DATA_TYPE], None],
+    ) -> None:
+        """Register a TTN_Entity handling for a platform.
+
+        New entries discovered by this coordinator will be checked with each registered platform.
+        If their `manages_uplink` method returns True then they will be added to their platform.
+        """
+
+        self.__entity_class_register.append(
+            TTNCoordinator.__RegisteredEntityClass(
+                self.__entry, entity_class, async_add_entities
+            )
+        )
+
+    def __async_add_entities(self, data: TTNClient.DATA_TYPE) -> None:
+        """Create new entities out of received TTN data if they are seen for the first time."""
+
+        for entity_class in self.__entity_class_register:
+            entity_class.async_add_entities(data)
+
+    class __RegisteredEntityClass:
+        def __init__(
+            self,
+            entry: ConfigEntry,
+            entity_class: type["TTN_Entity"],
+            async_add_entities: Callable[[TTNClient.DATA_TYPE], None],
+        ) -> None:
+            self.__entry = entry
+            self.__entity_class = entity_class
+            self.__async_add_entities = async_add_entities
+
+        def async_add_entities(self, data: TTNClient.DATA_TYPE) -> None:
+            """Add newly discovered entities to platform claiming ownership."""
+
+            entrySettings = TTN_EntrySettings(self.__entry)
+            self.__async_add_entities(
+                self.__entity_class(
+                    self.__entry,
+                    entrySettings.get_coordinator(),
+                    ttn_value,
+                )
+                for device_id, device_uplinks in data.items()
+                for field_id, ttn_value in device_uplinks.items()
+                if not self.__entity_class.exits(self.__entry, device_id, field_id)
+                and self.__entity_class.manages_uplink(entrySettings, ttn_value)
+            )
```

Comment:
I think it would be way cleaner to keep the coordinator responsible for fetching the data and have separate methods in the async setup entry in the platforms responsible for adding new entities. An example of this can be found at `withings`, especially the sensor platform.

With that approach it would create a listener, and I believe you mentioned something about not always polling, can you maybe give more information on when you exactly want to poll?